File: 2369-check-if-there-is-a-valid-partition-for-the-array/2369-check-if-there-is-a-valid-partition-for-the-array.py
class Solution:
    def validPartition(self, nums: List[int]) -> bool:
        dp = {-1 : True}
        def f(i):
            if i in dp: return dp[i]
            ans = False
            if i>0 and nums[i]==nums[i-1]: ans |= f(i-2)
            if i>1 and nums[i]==nums[i-1]==nums[i-2]: ans |= f(i-3)
            if i>1 and nums[i]==nums[i-1]+1==nums[i-2]+2: ans |= f(i-3) 
            dp[i] = ans
            return ans
        return f(len(nums)-1)

File: 2369-check-if-there-is-a-valid-partition-for-the-array/2369-check-if-there-is-a-valid-partition-for-the-array.cpp
class Solution {
public:
    bool validPartition(vector<int>& nums) {
        map<int, bool> dp;
        dp[-1] = true;
        return f(nums.size()-1, nums, dp);
    }
    bool f(int i, vector<int>&nums, map<int, bool> &dp){
        if(dp.find(i) != dp.end()) return dp[i];
        bool ans = false;
        if(i>0 && nums[i]==nums[i-1]) ans |= f(i-2, nums, dp);
        if(i>1 && nums[i]==nums[i-1] && nums[i-1]==nums[i-2]){
            ans |= f(i-3, nums, dp);
        }
        if(i>1 && nums[i]==nums[i-1]+1 && nums[i-1]==nums[i-2]+1){
            ans |= f(i-3, nums, dp);
        }
        dp[i] = ans;
        return ans;
    }
};